### PR TITLE
Create a struct for node metadata

### PR DESCRIPTION
**Pull request by howardjohn**
The goal here was to avoid doing our lookups into a `map[string]string` then doing hacky things like marshalling from a string, etc. We also ignore non-string fields right now, which puts us in a weird state where we can't read all metadata. 

One specific area where this hurt us in the past is that we made a new field POD_PORTS which is a json list, but marshaled as a list because we don't support list - this type of hack is essentially forever, as we need to support older versions, so now we will always have this oddity.

This does *not* make changes to the bootstrap logic yet. If we want this change I'll do that in a followup/add it to this PR

I thought this would be simple and a obvious improvement but it ended up being pretty complicated.. let me know what you guys think.

### CODE REVIEW

**Review by rshriram — COMMENTED** [review]
drive by comment:
convert node metadata into a proto serialize the proto into any and stick that into struct.
there are also potential pitfalls here though. unrelated info cant be sent by envoy

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go [hunk 1]
```diff
@@ -562,7 +562,7 @@ func proxyMatch(proxy *model.Proxy, cp *model.EnvoyFilterConfigPatchWrapper) boo
 	}
 
 	for k, v := range cp.Match.Proxy.Metadata {
-		if proxy.Metadata[k] != v {
+		if proxy.Metadata.Raw[k] != v {
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go]: hmm. i guess the any in struct wont work here.

**Review by rshriram — COMMENTED** [review]
looks good on first pass

File: pilot/pkg/model/context.go [hunk 2]
```diff
@@ -119,6 +120,182 @@ var (
 	istioVersionRegexp = regexp.MustCompile(`^([1-9]+)\.([0-9]+)(\.([0-9]+))?`)
 )
 
+// CsvList is a list that will be marshaled to a csv in Json
+type CsvList []string
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/context.go]: what is `csv`?

File: pilot/pkg/model/context.go [hunk 3]
```diff
@@ -119,6 +120,223 @@ var (
...
+		return err
+	}
+	*l = pl
+	return nil
+}
+
+// NodeMetadata defines the metadata associated with a proxy
+// Fields should not be assumed to exist on the proxy, especially newly added fields which will not exist
+// on older versions.
+type NodeMetadata struct {
```
Comment (douglas-reid) [inline pilot/pkg/model/context.go]: do we want a notice about not changing the `json` names and/or keeping them in sync with things like `EXCHANGE_KEYS` somehow?

File: pkg/bootstrap/bootstrap_config.go [hunk 4]
```diff
@@ -74,19 +73,18 @@ var (
...
-			model.NodeMetadataInstanceIPs,
-			model.NodeMetadataLabels,
-			model.NodeMetadataOwner,
-			model.NodeMetadataPlatformMetadata,
-			model.NodeMetadataWorkloadName,
-			model.NodeMetadataCanonicalTelemetryService,
-			model.NodeMetadataMeshID,
-			model.NodeMetadataServiceAccount,
-		}, ",")
+	metadataExchangeKeys = []string{
```
Comment (douglas-reid) [inline pkg/bootstrap/bootstrap_config.go]: do we care about keeping this in-sync with the `json` serializations from the struct?

**Review by douglas-reid — COMMENTED** [review]
looks reasonable to me.

File: pkg/bootstrap/bootstrap_config.go (hunk 4 above)
Comment (howardjohn) [reply pkg/bootstrap/bootstrap_config.go]: Good point. So as far as I can tell we cannot have a single const that declares both here and the json serialization (cant do something like ```name string fmt.Sprintf(`json:"%s"`, SomeConstant)```), maybe just a strongly worded comment? Open to suggestions here, this was very disappointing when I realized this

Comment (douglas-reid) [reply pkg/bootstrap/bootstrap_config.go]: I think a strongly-worded comment and some unit test are probably the best we can do.

**Review by hzxuzhonghu — APPROVED** [review]

**Review by rshriram — APPROVED** [review]

**Review by douglas-reid — APPROVED** [review]

**Review by JimmyCYJ — APPROVED** [review]

**Review by ayj — APPROVED** [review]

File: pilot/pkg/model/context.go [hunk 5]
```diff
@@ -119,6 +120,226 @@ var (
...
+	return nil
+}
+
+// NodeMetadata defines the metadata associated with a proxy
+// Fields should not be assumed to exist on the proxy, especially newly added fields which will not exist
+// on older versions.
+// The JSON field names should never change, as they are needed for backward compatibility with older proxies
+type NodeMetadata struct {
+	// IstioVersion specifies the Istio version associated with the proxy
+	IstioVersion string `json:"ISTIO_VERSION,omitempty"`
```
Comment (ramaraochavali) [inline pilot/pkg/model/context.go]: @howardjohn Older versions of proxies have the proxy version in "ISTIO_PROXY_VERSION" - was it changed to use "ISTIO_VERSION"?

Here is metdata dump of older (prior to 1.3)

```
     "metadata": {
      "INTERCEPTION_MODE": "REDIRECT",
      "CONFIG_NAMESPACE": "service-mesh",
      "WORKLOAD_NAME": "istio-ordering-66d468b7fb",
      "TLS_CLIENT_ROOT_CERT": "/client-certs/ca.pem",
      "sidecar.istio.io/inject": "true",
      "istio_app": "test",
      "madkub.sam.sfdc.net/allcerts": "{\"certreqs\":[{\"name\":\"tls-client-cert\",\"san\":null,\"role\":\"service-mesh.istio-ordering\",\"cert-type\":\"client\",\"kingdom\":\"prd\",\"superpod\":\"None\"},{\"name\":\"tls-server-cert\",\"san\":null,\"role\":\"service-mesh.istio-ordering\",\"cert-type\":\"server\",\"kingdom\":\"prd\",\"superpod\":\"None\"}]}",
      "POD_NAME": "istio-ordering-66d468b7fb-q5hnn",
      "istio": "sidecar",
      **"ISTIO_PROXY_VERSION": "1.1.3",**
      "sidecar.istio.io/statsInclusionPrefixes": "access_log_file,cluster,cluster_manager,control_plane,http,http2,http_mixer_filter,listener,listener_manager,redis,runtime,server,stats,tcp,tcp_mixer_filter,tracing",
      "INCLUDE_INBOUND_PORTS": "7443,15372",
      "NAMESPACE": "service-mesh",
      "EXCHANGE_KEYS": "NAME,NAMESPACE,INSTANCE_IPS,LABELS,OWNER,PLATFORM_METADATA,WORKLOAD_NAME,CANONICAL_TELEMETRY_SERVICE,MESH_ID,SERVICE_ACCOUNT",
      "TLS_SERVER_CERT_CHAIN": "/server-certs/server/certificates/server.pem",
      "INSTANCE_IPS": "10.251.130.102,fe80::42:aff:fefb:8266",
      "TLS_CLIENT_KEY": "/client-certs/client/keys/client-key.pem",
      "pod-template-hash": "2280246396",
      "POD_PORTS": "[\n    {\"name\":\"grpc-svc\",\"containerPort\":7443,\"protocol\":\"TCP\"}\n    ,{\"name\":\"scone-mgmt\",\"containerPort\":15372,\"protocol\":\"TCP\"}\n]",
      "TLS_SERVER_ROOT_CERT": "/server-certs/ca.pem",
      "SERVICE_ACCOUNT": "default",

      "TLS_CLIENT_CERT_CHAIN": "/client-certs/client/certificates/client.pem",
      "CLUSTER_ID": "Kubernetes",

}
```

**Comment by howardjohn** [comment]
yes new proxy do not have that field anymore

**Comment by ramaraochavali** [comment]
ok. Was it changed before 1.3? So the current (Master) pilot when working with older versions of Proxy, if it does not find that field,  it thinks it has max version https://github.com/istio/istio/blob/master/pilot/pkg/model/context.go#L609 and sends all the latest config, which creates problems.

**Comment by ramaraochavali** [comment]
@howardjohn added a warning log line for this scenario https://github.com/istio/istio/pull/17492 - See if it makes sense

**Comment by howardjohn** [comment]
older proxy should have both fields ?

**Comment by ramaraochavali** [comment]
No. I do not find "ISTIO_VERSION" field in old proxy. Here is the metadata of older proxy

```
    "node": {
     "id": "sidecar~10.251.130.102~istio-ordering-66d468b7fb-q5hnn.service-mesh~service-mesh.svc.cluster.local",
     "cluster": "istio-ordering.service-mesh",
     "metadata": {
      "INTERCEPTION_MODE": "REDIRECT",
      "CONFIG_NAMESPACE": "service-mesh",
      "WORKLOAD_NAME": "istio-ordering-66d468b7fb",
      "TLS_CLIENT_ROOT_CERT": "/client-certs/ca.pem",
      "sidecar.istio.io/inject": "true",
      "istio_app": "test",
      "madkub.sam.sfdc.net/allcerts": "{\"certreqs\":[{\"name\":\"tls-client-cert\",\"san\":null,\"role\":\"service-mesh.istio-ordering\",\"cert-type\":\"client\",\"kingdom\":\"prd\",\"superpod\":\"None\"},{\"name\":\"tls-server-cert\",\"san\":null,\"role\":\"service-mesh.istio-ordering\",\"cert-type\":\"server\",\"kingdom\":\"prd\",\"superpod\":\"None\"}]}",
      "POD_NAME": "istio-ordering-66d468b7fb-q5hnn",
      "istio": "sidecar",
      "ISTIO_PROXY_VERSION": "1.1.3",
      "sidecar.istio.io/statsInclusionPrefixes": "access_log_file,cluster,cluster_manager,control_plane,http,http2,http_mixer_filter,listener,listener_manager,redis,runtime,server,stats,tcp,tcp_mixer_filter,tracing",
      "app": "istio-ordering",
      "INCLUDE_INBOUND_PORTS": "7443,15372",
      "NAMESPACE": "service-mesh",
      "EXCHANGE_KEYS": "NAME,NAMESPACE,INSTANCE_IPS,LABELS,OWNER,PLATFORM_METADATA,WORKLOAD_NAME,CANONICAL_TELEMETRY_SERVICE,MESH_ID,SERVICE_ACCOUNT",
      "TLS_SERVER_CERT_CHAIN": "/server-certs/server/certificates/server.pem",
      "INSTANCE_IPS": "10.251.130.102,fe80::42:aff:fefb:8266",
      "TLS_CLIENT_KEY": "/client-certs/client/keys/client-key.pem",
      "pod-template-hash": "2280246396",
      "POD_PORTS": "[\n    {\"name\":\"grpc-svc\",\"containerPort\":7443,\"protocol\":\"TCP\"}\n    ,{\"name\":\"scone-mgmt\",\"containerPort\":15372,\"protocol\":\"TCP\"}\n]",
      "TLS_SERVER_ROOT_CERT": "/server-certs/ca.pem",
      "SERVICE_ACCOUNT": "default",
      "sam_function": "istio-ordering",
      "OWNER": "kubernetes://api/extensions/v1beta1/namespaces/service-mesh/replicasets/istio-ordering-66d468b7fb",
      "TLS_CLIENT_CERT_CHAIN": "/client-certs/client/certificates/client.pem",
      "hostname": "istio-ordering-66d468b7fb-q5hnn",
      "CLUSTER_ID": "Kubernetes",
      "LABELS": {
       "sam_function": "istio-ordering",
       "app": "istio-ordering",
       "sam_app": "istio-ordering",
       "istio_app": "test",
       "settings_path": "istio.-.prd.-.istio-ordering",
       "pod-template-hash": "2280246396"
      },
      "TLS_SERVER_KEY": "/server-certs/server/keys/server-key.pem",
      "NAME": "istio-ordering-66d468b7fb-q5hnn",
      "kubernetes_cluster_name": "",
      "namespace": "service-mesh",
      "sam_app": "istio-ordering",
      "settings_path": "istio.-.prd.-.istio-ordering"
     },
     "locality": {},
     "build_version": "7f57dc2f2755646d45a3652c3629c8afb1f8ea69/1.12.0-dev/Clean/RELEASE/BoringSSL"
    },
```

**Comment by howardjohn** [comment]
what version is that? I think it's been around since at least 1.2. Master
pilot is certainly not compatible with envoy 1.1

**Comment by ramaraochavali** [comment]
@howardjohn This was a problem with our internal build pipeline that pushed wrong version indicator in metadata. We are fixing it. If you think this warning https://github.com/istio/istio/pull/17492 is not required, I will close the PR. Sorry for the trouble.

**Pull request merged**